README.md
# generator-combinator
Provides [parser-combinator](https://en.wikipedia.org/wiki/Parser_combinator)-like combinable text _generation_ in Rust.

You can add this crate to your Rust project with `generator-combinator = "0.4.0"`. [Documentation on docs.rs](https://docs.rs/generator-combinator) and [crates.io listing](https://crates.io/crates/generator-combinator).

To generate street address-like input, only a few components are required. We can quickly produce a range of nearly 1B possible values that can be fully iterated over or randomly sampled:

```rust
use generator_combinator::Generator;
let space = Generator::from(' ');

// 3-5 digits for the street number. If the generated value has leading 0s, trim them out
let number = (Generator::Digit * (3, 5)).transform(|s| {
    if s.starts_with('0') {
        s.trim_start_matches('0').to_string()
    } else {
        s
    }
});


let directional = space.clone() + oneof!("N", "E", "S", "W", "NE", "SE", "SW", "NW");
let street_names = space.clone() + oneof!("Boren", "Olive", "Spring", "Cherry", "Seneca", "Yesler", "Madison", "James", "Union", "Mercer");
let street_suffixes = space.clone() + oneof!("Rd", "St", "Ave", "Blvd", "Ln", "Dr", "Way", "Ct", "Pl");

let address = number
    + directional.clone().optional() // optional pre-directional
    + street_names
    + street_suffixes
    + directional.clone().optional(); // optional post-directional

assert_eq!(address.len(), 809_190_000);

// With the 'with_rand' feature:
let addr_values = address.values();
println!("Example: {}", addr_values.random()); //Example: 344 W Yesler Way
println!("Example: {}", addr_values.random()); //Example: 702 NE Spring Ct N
println!("Example: {}", addr_values.random()); //Example: 803 SW Madison Way SE
```

This library is 0.4.0 - there may be issues, functionality may be incomplete, etc. 

## Known issues / _nota bene_
- Generated digits include leading zeros. Use `.transform` to address this, if desired.

## TODO
- [x] Consider including `Fn` variants of `Generator`
- [x] `Generator` post-processing of component strings (eg, to strip leading zeros) before combining output

## License

Licensed under either of

 * Apache License, Version 2.0, ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)

at your option.

### Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted for inclusion in the work by you, as defined in the Apache-2.0 license, shall be dual licensed as above, without any additional terms or conditions.


CHANGELOG.md
## [0.4.0] - 2022-04-16
### Changed
- Made `rand` optional by default; use `features = ["with_rand"]` to enable the `random` function.
- `visit_one` now passes ownership of the `String` instead of a `&str`. It doesn't need to maintain ownership, and if the caller wants it, this may avoid unnecessary clones.
- Changes for Clippy.

### Added
- `impl From<String> for Generator`: you could previously say `let g : Generator = "hello".into();` for `&str` and can now do so with a `String`
- `Add<T>`, `AddAssign<T>`, `BitOr<T>`, and `BitOrAssign<T>` implementations for `char`, `&str`, and `String`; for example:
   ```rust
   let mut g = Generator::from("hello");
   g |= "salut";
   g += ',';
   g += " ";
   g += Generator::from("world") | "tout le monde" | "🌎";
   g += String::from("!");
   ```
- Extra test for a pattern that exceeds `u128` capacity.
- Two example programs in the `/examples/` folder.
- `Generator::Empty` variant that is basically a no-op and is also the value returned by the new `impl Default for Generator`.

### Fixed
- `BitOrAssign` for two OneOf variants incorrectly included an extra digit:
   ```rust
   let mut g = oneof!('a', 'b');
   g += oneof!('x', 'y');
   ```

## [0.3.0] - 2021-06-05
### Changed
- Renamed `Generator::generate_exact` to `Generator::generate_one` for generating a single string from a Generator
- Renamed `Generator::values` to `Generator::generate_all` for iterating over a Generator's range
- Renamed `Iter` to `IterString` for string iteration over a Generator's range, returned by `Generator::generator_all`

### Added
- `Generator::visit_one` to provide each part, in order, as a `&str`. For cases where the entire string isn't necessarily needed at once or at all, this avoids allocating and freeing memory.
- [quickcheck](https://github.com/BurntSushi/quickcheck) testing
- Additional tests, including of `Generator::regex`

src/lib.rs
//! Generates [`String`] combinations from combinable [`Generator`]s.
//!
//! `generator-combinator` aims to provide text _generation_ capabilities, similar to but in the opposite direction of [parser combinators](https://en.wikipedia.org/wiki/Parser_combinator) that transform text into structured data.
//!
//! # Iris example
//! Consider the [regular expression](https://en.wikipedia.org/wiki/Regular_expression) `iris( (versicolor|virginica|setosa))?`. This regex matches exactly four input values:
//! * `"iris"`
//! * `"iris versicolor"`
//! * `"iris virginica"`
//! * `"iris setosa"`
//!
//! This regex does _not_ match other values such as `"iris fulva"` or `"iris "` (with trailing space). If we want to generate these four valid input values (something like a reverse regex), we can build a generator-combinator as:

//! ```
//! use generator_combinator::{oneof, Generator};
//! let genus = Generator::from("iris");
//!
//! // Support three different species
//! let species = Generator::from("versicolor")
//!    | Generator::from("virginica")
//!    | Generator::from("setosa");
//! // Alternately:
//! let species = oneof!("versicolor", "virginica", "setosa");
//!
//! // Delimit the genus and species with a space
//! let species = Generator::from(' ') + species;
//!
//! // Allow generated values to be genus-only or with the species
//! let iris = genus + species.optional();
//!
//! // Our generator should produce exactly four values
//! assert_eq!(iris.len(), 4);
//!
//! let mut iris_values = iris.generate_all();
//! assert_eq!(iris_values.next(), Some("iris".into()));
//! assert_eq!(iris_values.next(), Some("iris versicolor".into()));
//! assert_eq!(iris_values.next(), Some("iris virginica".into()));
//! assert_eq!(iris_values.next(), Some("iris setosa".into()));
//! assert_eq!(iris_values.next(), None);
//!
//! assert_eq!(iris.regex(), "iris( (versicolor|virginica|setosa))?");
//! ```
//!
//! # Street address example
//! Generators can be used to produce sample input data according to some pattern. For example, to generate street addresses (which aren't necessarily verifiable):
//! ```
//! use generator_combinator::{Generator, oneof, gen};
//! let space = Generator::from(' ');
//!
//! let number = (Generator::Digit * (3, 5)).transform(|s| s.trim_start_matches('0').to_string());
//!
//! let directional = space.clone() + oneof!("N", "E", "S", "W", "NE", "SE", "SW", "NW");
//! let street_names = space.clone() + oneof!("Boren", "Olive", "Spring", "Cherry", "Seneca", "Yesler", "Madison", "James", "Union", "Mercer");
//! let street_suffixes = space.clone() + oneof!("Rd", "St", "Ave", "Blvd", "Ln", "Dr", "Way", "Ct", "Pl");
//!
//! let address = number
//!     + directional.clone().optional()
//!     + street_names
//!     + street_suffixes
//!     + directional.clone().optional();
//!
//! assert_eq!(address.len(), 809_190_000);
//!
//! #[cfg(feature = "with_rand")]
//! {
//!     let addr_values = address.generate_all();
//!     println!("Example: {}", addr_values.random()); //Example: 344 W Yesler Way
//!     println!("Example: {}", addr_values.random()); //Example: 702 NE Spring Ct N
//!     println!("Example: {}", addr_values.random()); //Example: 803 SW Madison Way SE
//! }

#[cfg(test)]
#[macro_use]
extern crate quickcheck;

mod macros;

mod generator;
pub use generator::Generator;

mod iter;
pub use iter::StringIter;

mod transformfn;
